### DISCUSSION TITLE
Expose a webservice

### DISCUSSION THREAD

**Discussion by OleJ1964** [hlky/stable-diffusion-webui#875] [Feature suggestions]
Would it be possible to expose the SD as a webservice?

**Comment by outlookhazy** [comment]
I spent some time reverse engineering the gradio api that the UI is using - it appears to just be RPC over websockets. Biggest issue with that was that it appears likely auto-generated (referring to functions by number) so it wouldn't be stable as the source changes.

**Reply by OleJ1964** [reply]
Thanks for reply :-) I understand - but still better than spredning implementations all-over the place - id really like to be able to implement SD in various applications/projects (languages/platforms) - and a webservice just seem right (at the moment anyway - who knows where we are in 1 month)

**Reply by outlookhazy** [reply]
I agree entirely.
I _would_ just build my own server that calls the python scripts as an external process, _but_ it seems like a good portion of time is spent loading models, so I would guess that doing it that way is going to be much slower than having a python server managing what's loaded at any given time. Unfortunately, I don't have any background in Python, so I would have to start from the ground up.

**Reply by endomorphosis** [reply]
I added a pull request, which only modifies 3 lines, which exposes an API through gradio.

**Reply by outlookhazy** [reply]
Thanks for this! It seems the API is not well documented at all.. 
I'm having trouble finding any substantial information.
I found a post that suggests adding that api_name tag adds api endpoints in the form \<host>:\<port>/api/\<api_name>, but I haven't been able to find exactly what that endpoint **is** (http? websocket?) what it's expecting (json? content?) or what return data would look like.. lol

Edit: haven't played with it yet, but the gradio tests _seem_ to indicate the api endpoint is http. It's expecting a POST containing a json object in the format `{"data":["param1","param2"]}` and returns in the same format

**Reply by endomorphosis** [reply]
my pull request has been accepted, and you are correct, that is the format that it expects, which is the same format as is being sent over websockets

**Comment by OleJ1964** [comment]
Exactly - that was my initial thought. but the time increase I feel would
not merrit it. If I knew anything about python I'd look into it myself. But
i plan looking into python soon just bc. SD. (I really don't like the
language or environment)

Den fre. 9. sep. 2022 kl. 18.13 skrev Charlie ***@***.***>:

**Comment by OleJ1964** [comment]
Sounds absolutely fantastic.

**Comment by RaigaK** [comment]
@endomorphosis can you please share an example ?
Cant get it to work, response is always `{"error":"'NoneType' object is not callable"}`
im following the view api page of Gradio

**Reply by endomorphosis** [reply]
i think that the package maintainers have broken it, it changed so frequently and we go so annoyed with the moving target, that me and my partner just reimplemented our own api solution.

a previous pull request that I had, does in fact work appropriately, but there are multiple stages of calling the api to add a job, then checking on that queue until the response is ready, then finally reading the response.
